### DISCUSSION TITLE
Interfacing between Nucleo(STM32F446RET) and A7672s (GSM Module)

### DISCUSSION THREAD

**Discussion by Nikitanagar** [stm32duino/Arduino_Core_STM32#2593] [Q&A] [answered]
I am working on interfacing an A7672S GSM module with an STM32F446RET microcontroller. I am using the STM32duino core, downloaded from the official [STM32duino BoardManager Files](https://github.com/stm32duino/BoardManagerFiles/raw/main/package_stmicroelectronics_index.json). However, the GSM module is not responding to AT commands, and I don’t see any output on the Serial Monitor.

**Hardware Setup:** 
Microcontroller: STM32F446RET
GSM Module: A7672S
Connections:
PA2 (TX) -> RX (GSM)
PA3 (RX) -> TX (GSM)
GND (STM32) -> GND (GSM)
Power supply for the GSM module is sufficient (verified). 

**Code:**
Below is the code I am using:
```C++
HardwareSerial Serial2(PA3, PA2);

#define gsm Serial2

void setup()
{
  gsm.begin(115200);   // Setting the baud rate of GSM Module
  Serial.begin(115200);    // Setting the baud rate of Serial Monitor (Arduino)
  delay(100);
}


void loop()
{
  if (Serial.available() > 0)
    switch (Serial.read())
    {
      case 's':
        SendMessage();
        break;
      case 'r':
        RecieveMessage();
        break;
    }

  if (gsm.available() > 0)
    Serial.write(gsm.read());
}


void SendMessage()
{
  gsm.println("AT");
  gsm.println("AT+CMGF=1");    //Sets the GSM Module in Text Mode
  delay(1000);  // Delay of 1000 milli seconds or 1 second
  gsm.println("AT+CMGS=\"+910000000000\"\r"); // Replace x with mobile number
  delay(1000);
  gsm.println("I am SMS from GSM Module");// The SMS text you want to send
  delay(100);
  gsm.println((char)26);// ASCII code of CTRL+Z
  delay(1000);
}

void RecieveMessage()
{
  gsm.read();
  gsm.println("AT+CNMI=2,2,0,0,0"); // AT Command to receive a live SMS
  delay(1000);
}
```

**Expected Behavior:**
When I press 's' or 'r' in the Serial Monitor:
The GSM module should respond to AT commands.
SMS should be sent or received depending on the command.

**Observed Behavior:**
No response is observed in the Serial Monitor.
The GSM module appears unresponsive to AT commands.

**Debugging Attempts:**
- Verified the UART connections (TX ↔ RX).
- Tested the GSM module independently using a USB-to-serial adapter and verified it works.
- Ensured the GSM module has sufficient power supply.
- Checked the STM32 clock configuration in the Arduino IDE:
- Tools > U(S)ART Support: Enabled.

**Questions:** 
1. Is the HardwareSerial configuration for Serial2 correct?
2. Could the baud rate mismatch be causing the issue? (The GSM module defaults to 115200.)
3. Is there any additional configuration required for the STM32F446RET to use HardwareSerial?
4. Are there any known issues with the STM32duino core for STM32F446 and GSM module communication?

**Additional Information:**
- Arduino IDE version: 2.3.4
- STM32duino core version: 2.9.0
- GSM module AT commands work correctly with other microcontrollers (tested with ESP32).

**Comment by fpistm** [comment]
Hi @Nikitanagar 
Which board you used ? a custom board and then you used the generic target? Generic F446RETx?

I don't understand this:

> Tools > CPU Speed: Set to 180 MHz.

This core does not provide such option. Are you sure you used this core?

**Reply by Nikitanagar** [reply]
yeah you are right, It was the mistake. I am using [Nucleo32](https://os.mbed.com/platforms/ST-Nucleo-F446RE/) which consist the STm32F446RET chip.

**Comment by fpistm** [comment]
It is a Nucleo 64 not 32.
So you have to Select the correct board in the menu:
Tools -> Boards: Nucleo-64
Then Tools -> Board part number: Nucleo F446RE

Your sketch will not build as the default `Serial` instance is mapped to `Serial2` using `PA3/PA2`.
https://github.com/stm32duino/Arduino_Core_STM32/blob/ffb23d6b748001dfd3f8111d37610bb51022bc81/variants/STM32F4xx/F446R(C-E)T/variant_NUCLEO_F446RE.h#L148-L160

But your sketch use those pins for the GSM. So you have to instantiate an other one.

**Comment by fpistm** [comment]
Pay attention the pin D0 and D1 of the Arduino Uno connector are not connected to the pins PA2/PA3:
![image](https://github.com/user-attachments/assets/6c16fd5f-13cb-46d0-933d-c15cac3e3876)

![image](https://github.com/user-attachments/assets/876fa933-d981-4e4c-a400-a5c6c3b0c620)
Those pins are connected to the STLink only .
You can use PA9 (D8) and PA10 (D2) for the GSM:

`HardwareSerial gsm(PA10, PA9);`

**Reply by Nikitanagar** [reply]
Thank you @fpistm for this crucial information, although I was aware about pinouts, but I was unaware of the STlink pins. Thank you
Now I replaced the pins from PA2, PA3 to  PA9, PA10 

Here is the updated code: 

HardwareSerial gsm(PA10, PA9);

void setup()
{
  gsm.begin(115200);   // Setting the baud rate of GSM Module
  Serial.begin(115200);    // Setting the baud rate of Serial Monitor (Arduino)
  delay(100);

  gsm.print("AT");
  if (gsm.available() > 0){
    gsm.write(gsm.read());
  }
}


void loop()
{
  if (Serial.available() > 0)
    switch (Serial.read())
    {
      case 's':
        SendMessage();
        break;
      case 'r':
        RecieveMessage();
        break;
    }

  if (gsm.available() > 0)
    gsm.write(gsm.read());
}


void SendMessage()
{
  gsm.println("AT");
  gsm.println("AT+CMGF=1");    //Sets the GSM Module in Text Mode
  delay(1000);  // Delay of 1000 milli seconds or 1 second
  gsm.println("AT+CMGS=\"+910000000000\"\r"); // Replace x with mobile number
  delay(1000);
  gsm.println("I am SMS from GSM Module");// The SMS text you want to send
  delay(100);
  gsm.println((char)26);// ASCII code of CTRL+Z
  delay(1000);
}

void RecieveMessage()
{
  if (gsm.available() > 0){
    gsm.write(gsm.read());
  }

  gsm.println("AT+CNMI=2,2,0,0,0"); // AT Command to receive a live SMS
  delay(1000);
}`
is there any thing I should have to improve, because issue is still same.

**Reply by fpistm** [reply]
I would remove the delay and replace it by Serial.flush(); or gsm.flush();

Or kept the delay if required by the gsm module andasd the flush.

**Reply by Nikitanagar** [reply]
Thank you for your help, I wrote the code to test the AT commands.

```C++
HardwareSerial serialmon(PA3, PA2); // Serial monitor (connected to ST-Link)
HardwareSerial gsm(PA10, PA9);      // GSM module (connected to PA10, PA9)

void setup() {
  // Initialize both Serial interfaces
  serialmon.begin(115200);
  gsm.begin(115200);

  // Wait for serial monitor to be ready
  while (!serialmon) {
    // Do nothing, wait for connection
  }

  serialmon.println("Testing GSM module...");
}

void loop() {
  if (serialmon.available()) {
    String command = serialmon.readStringUntil('\n');
    gsm.println(command); 
    serialmon.println("Sent to GSM: " + command);
    delay(1000); // Wait for GSM to process the command

    if (gsm.available()) {
      String response = gsm.readString(); 
      serialmon.println("Response from GSM: " + response);
    } else {
      serialmon.println("No response from GSM.");
    }
  }
}
```

I hope this will help to someone.

**Reply by fpistm** [reply]
No need to define `serialmon`, default `Serial` instance uses PA2/PA3, so code become:

```C++
HardwareSerial gsm(PA10, PA9);      // GSM module (connected to PA10, PA9)

void setup() {
  // Initialize both Serial interfaces
  Serial.begin(115200);
  gsm.begin(115200);

  // Wait for serial monitor to be ready
  while (!Serial) {
    // Do nothing, wait for connection
  }

  Serial.println("Testing GSM module...");
}

void loop() {
  if (Serial.available()) {
    String command = Serial.readStringUntil('\n');
    gsm.println(command); 
    Serial.println("Sent to GSM: " + command);
    delay(1000); // Wait for GSM to process the command

    if (gsm.available()) {
      String response = gsm.readString(); 
      Serial.println("Response from GSM: " + response);
    } else {
      Serial.println("No response from GSM.");
    }
  }
}
```

**Reply by Nikitanagar** [reply]
Yeah i did that, but serial was not working, so that I mentioned the serialmon.

**Reply by fpistm** [reply]
Serial works except if you disabled it thanks the menu. (Enabled (no generic Serial))

**Reply by Nikitanagar** [reply]
Yeah I configured the settings properly, then I did this. You are right, but it's not working. Don't know the proper issue, but yeah it's not working without mentioning the serialmon.

**Reply by fpistm** [reply]
With default option Serial works as expected. Hard to tell what is your issue.
Could you do a screenshot of the options?

**Reply by Nikitanagar** [reply]
yeah Here is the attached image:
![Settings](https://github.com/user-attachments/assets/dd8b1b3b-86b2-4fce-8144-41e49076a802)

**Reply by fpistm** [reply, marked as answer]
Ok.. disable the USB support then Serial will be on the USART...
